src/main/scala/com/wavesplatform/matcher/model/OrderHistory.scala

```diff
@@ -49,22 +50,29 @@ class OrderHistory(db: DB, settings: MatcherSettings) {
       )
   }
 
-  private def saveOrderInfo(rw: RW, event: Event): Map[ByteStr, OrderInfoChange] =
-    saveOrderInfoTimer.measure(db.readWrite { rw =>
-      val orderInfoDiffs = collectChanges(event)
-
-      val updatedInfo = orderInfoDiffs.map {
-        case (order, orderInfoDiff) =>
-          val orderId  = order.id()
-          val orig     = DBUtils.orderInfoOpt(rw, orderId)
-          val combined = combine(order, orig, orderInfoDiff)
-          val change   = OrderInfoChange(order, orig, combined)
-          rw.put(MatcherKeys.orderInfo(orderId), change.updatedInfo)
-          orderId -> change
-      }
+  private def saveOrderInfo(rw: RW, event: Event): Unit = saveOrderInfoTimer.measure {
+    val orderInfoDiffs = collectChanges(event)
+
+    val (_, changes) = orderInfoDiffs.foldLeft((Map.empty: ChangedKeys, Map.empty[Order.Id, OrderInfoChange])) {
+      case ((origChangedKeys, origChanges), (order, orderInfoDiff)) =>
+        val orderId      = order.id()
+        val origInfo     = DBUtils.orderInfoOpt(rw, orderId)
+        val combinedInfo = combine(order, origInfo, orderInfoDiff)
+        val change       = OrderInfoChange(order, origInfo, combinedInfo)
+        rw.put(MatcherKeys.orderInfo(orderId), change.updatedInfo)
 
-      updatedInfo.toMap
-    })
+        val changedKeys1 = origChangedKeys.updated(MatcherKeys.orderInfo(orderId), change.updatedInfo)
+        val changedKeys2 = if (origInfo.isEmpty) {
+          saveOrder(rw, order)
+          addOrderIndexes(rw, change, changedKeys1)
+        } else changedKeys1
+
+        (updateOldestActiveNr(rw, change, changedKeys2), origChanges.updated(orderId, change))
+    }
+
+    val opDiff = diff(event, changes)
+    saveOpenVolume(rw, opDiff)
```

Comment:
Just realised, I can make saveOpenVolume work with (Address, OpenPortfolio), not the Map, and then able to simplify this method. But this changes will be later in a bug-fix task :)